R/name_usage.r

```diff
@@ -81,21 +84,63 @@ name_usage <- function(key=NULL, name=NULL, data='all', language=NULL,
     stop("Parameters not currently accepted: \n sourceId")
   }
 
+  check_vals(limit, "limit")
+
   # each of these args must be length=1
   if (!is.null(rank)) stopifnot(length(rank) == 1)
   if (!is.null(name)) stopifnot(length(name) == 1)
   if (!is.null(language)) stopifnot(length(language) == 1)
   if (!is.null(datasetKey)) stopifnot(length(datasetKey) == 1)
 
   args <- rgbif_compact(list(offset = start, limit = limit,
-                             sourceId = sourceId))
+                             sourceId = sourceId,
+                             datasetKey = datasetKey))
   args <- c(args, rank, datasetKey, name, language)
   data <- match.arg(data,
       choices = c('all', 'verbatim', 'name', 'parents', 'children',
                 'related', 'synonyms', 'descriptions',
                 'distributions', 'media', 'references', 'speciesProfiles',
                 'vernacularNames', 'typeSpecimens', 'root'), several.ok = FALSE)
-  out <- getdata(data, key, uuid, shortname, args, curlopts)
+  # paging implementation
+  if (limit > 1000) {
+    if (!("offset" %in% names(args)))
+      args$offset <- 0
+    iter <- 0
+    sumreturned <- 0
+    numreturned <- 0
+    outout <- list()
+    while (sumreturned < limit) {
+      iter <- iter + 1
+      tt <- getdata(data, key, uuid, shortname, args, curlopts)
+      # if no results, assign numreturned var with 0
+      if (identical(tt$results, list())) {
+        numreturned <- 0}
+      else {
+        numreturned <- length(tt$results)}
+      sumreturned <- sumreturned + numreturned
+      # if less results than maximum
+      if ((numreturned > 0) & (numreturned < 1000)) {
+        # update limit for metadata before exiting
+        limit <- numreturned
+        args$limit <- limit
+      }
+      if (sumreturned < limit) {
+        # update args for next query
+        args$offset <- args$offset + numreturned
+        args$limit <- limit - sumreturned
+      }
+      outout[[iter]] <- tt
+    }
+    out <- list()
+    out$results <- purrr::map(outout, ~purrr::splice(.$results))
```

Comment:
i don't want to add a new pkg dependency unless there is very good reason to do so. Is the `purrrr` code doing something like `lapply(outout, "[[", "results")` ?

Reply:
Done. e88243a.
i *copied* your way of merging `.results` in `occ_search`. It works perfectly. So, purrr is removed.